app/src/main/java/org/mozilla/tv/firefox/pocket/PocketViewModel.kt

```diff
@@ -0,0 +1,55 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox.pocket
+
+import android.arch.lifecycle.MutableLiveData
+import android.arch.lifecycle.ViewModel
+
+sealed class PocketViewModelState {
```

Comment:
Observing LiveData is pretty noisy, and we're generally going to be observing all of it in the same place anyway, so I'm thinking VMs should wrap all of their state into a single object and just emit that.  In this case the state is either a list of tiles or error, and I'm not sure what else I would call that.  I'm open to suggestions, but I think this is appropriate.

I do agree on moving inside the VM for namespacing, and I'll shorten the name accordingly.

Reply:
> Observing LiveData is pretty noisy,

What do you mean?

> and we're generally going to be observing all of it in the same place anyway, so I'm thinking VMs should wrap all of their state into a single object and just emit that.

This is purely hypothetical at the moment because Pocket only has one state (Loading), but I think have one global state loses the advantages (and disadvantages) of encapsulation: isolating unrelated state and their transitions so that the smaller amount of code is easier to understand and making it more difficult to allow state to unintentionally couple.

In the short term, I'd prefer to be specific and name it after the one thing it is - a loading state - so it's clearer to someone what exactly code manipulated the `PocketViewModelState` entails. If someone expands that object to encapsulate more state because they feel it's a better design, then it can take a more general name.

Reply:
>> Observing LiveData is pretty noisy,

> What do you mean?

There's enough boilerplate around it that if, for example, the VM carried three different values, I wouldn't want each to have its own LiveData.
```kotlin
personVm.name.observe(fragment.viewLifecycleOwner, Observer { name ->
  nameTv.text = name
}
personVm.age.observe(fragment.viewLifecycleOwner, Observer { age ->
  ageTv.text = age
}
personVm.height.observe(fragment.viewLifecycleOwner, Observer { height ->
  heightTv.text = height
}
```
as opposed to
```kotlin
personVm.state.observe(fragment.viewLifecycleOwner, Observer { state ->
  setName(state.name)
  setAge(state.age)
  setHeight(state.height)
}
```

> Pocket only has one state (Loading)

Tbh, I'm not sure I would agree with this assessment.  As is, Pocket VM state covers:
- Loading vs complete
- Error state
- Feed contents (images, urls, etc)
- Feed is clickable vs not

I don't feel that `LoadingState` properly conveys that.

> I think have one global state loses the advantages (and disadvantages) of encapsulation

True, but imo we're encapsulating state by breaking it up into relatively small view models.  I'm not sure there's much of a difference between organizing the interiors of the VMs to contain several independent LiveDatas and having one LiveData that delivers all of the state.  If we see any VMs that are managing too much state, I think the prefered solution would be to break them into smaller, composable VMs wherever possible.

Reply:
> If we see any VMs that are managing too much state, I think the prefered solution would be to break them into smaller, composable VMs wherever possible.

My concern with this is that we lose specificity on the naming of the state objects, which makes it hard to understand what the object does without looking at its definition, hurting abstraction ability.

I think the naming/encapsulation trade-offs seems to be on a case-by-case basis so I don't feel too strongly here.